src/extensions/extensionsHelper.ts

```diff
@@ -479,51 +568,78 @@ export async function incrementPrereleaseVersion(
+  extensionRef: string,
+  newVersion: string
+): Promise<{ versionByStage: Map<string, string>; hasVersions: boolean }> {
+  const versionByStage = new Map(
+    ["rc", "alpha", "beta"].map((stage) => [
+      stage,
+      semver.inc(`${newVersion}-${stage}`, "prerelease", undefined, stage)!,
+    ])
+  );
+  let extensionVersions: ExtensionVersion[] = [];
+  try {
+    extensionVersions = await listExtensionVersions(extensionRef, `id="${newVersion}"`, true);
+    for (const extensionVersion of extensionVersions) {
+      const version = semver.parse(extensionVersion.spec.version)!;
+      // Extensions only support semvers of format: <major>.<minor>.<patch>-<stage>.<count>
+      const stage = String(version.prerelease[0]).split(".")[0];
+      if (versionByStage.has(stage) && semver.gte(version, versionByStage.get(stage)!)) {
+        versionByStage.set(stage, semver.inc(version, "prerelease", undefined, stage)!);
+      }
+    }
+  } catch (e) {
+    // Fallthrough.
+  }
+  versionByStage.set("stable", newVersion);
+  return { versionByStage, hasVersions: extensionVersions.length > 0 };
+}
+
+/**
+ * Validates the extension spec.
  *
+ * @param rootDirectory the directory with the extension source
+ * @param extensionRef the ref of the extension
  */
-async function validateExtensionSpec(args: {
-  publisherId: string;
-  extensionId: string;
-  rootDirectory: string;
-  latestVersion?: string;
-  stage: ReleaseStage;
-}): Promise<{ extensionSpec: ExtensionSpec; notes: string }> {
-  const extensionRef = `${args.publisherId}/${args.extensionId}`;
-  const extensionSpec = await getLocalExtensionSpec(args.rootDirectory);
-  if (extensionSpec.name !== args.extensionId) {
+async function validateExtensionSpec(
+  rootDirectory: string,
+  extensionId: string
+): Promise<ExtensionSpec> {
+  const extensionSpec = await getLocalExtensionSpec(rootDirectory);
+  if (extensionSpec.name !== extensionId) {
     throw new FirebaseError(
       `Extension ID '${clc.bold(
-        args.extensionId
+        extensionId
       )}' does not match the name in extension.yaml '${clc.bold(extensionSpec.name)}'.`
     );
   }
-
   // Substitute deepcopied spec with autopopulated params, and make sure that it passes basic extension.yaml validation.
   const subbedSpec = JSON.parse(JSON.stringify(extensionSpec));
   subbedSpec.params = substituteParams<Param[]>(
     extensionSpec.params || [],
     AUTOPOULATED_PARAM_PLACEHOLDERS
   );
   validateSpec(subbedSpec);
+  return extensionSpec;
+}
 
-  // Increment the pre-release annotation if it is not a stable version.
-  extensionSpec.version = await incrementPrereleaseVersion(
-    extensionRef,
-    extensionSpec.version,
-    args.stage
-  );
-
+/**
+ * Validates the release notes.
+ *
+ * @param rootDirectory the directory with the extension source
+ * @param newVersion the new extension version
+ */
+function validateReleaseNotes(rootDirectory: string, newVersion: string, required: boolean) {
```

Comment:
I'd prefer not to add 'required' as an argument to this function - it spreads the logic of when release notes are necessary across the codebase instead of keeping confined to a single function.